### ISSUE TITLE
importlib.import_module load module twice

### ISSUE THREAD

**Issue by Hazhzeng** [Azure/azure-functions-python-worker#752]
### Investigative information
Thanks Leighton for reporting this issue.
https://github.com/Azure/azure-functions-python-worker/blob/dev/azure_functions_worker/loader.py#L76 The **importlib.import_module** will load function definition twice in customer's code

```python
import azure.functions as func

print('LOAD_TWICE')

def main(req: func.HttpRequest):
    return 'OK'
```
Will results in:
LOAD_TWICE
LOAD_TWICE

This is a Python by designed behavior, since the \_\_init\_\_.py file will be executed twice.
1. The first time is when the Trigger package and any sub-package is first imported, \_\_init\_\_.py  file will be executed.
2. The second time is to explicitly load the code inside \_\_init\_\_.py file.

When the function hits

- Commit: 8f91e63bf7c46a28dade08947c1ef65b49b92be2
- Function App name: local app
- Core Tools version: 3.0.2931

### Repro steps
1. Create a simple HTTP trigger
2. Write a print statement outside the def main(): scope
3. The print statement will execute twice

### Expected behavior
The print statement should only execute once

### Actual behavior
The print statement executes twice

### Known workarounds
1. **Change function.json => scriptFile from "\_\_init\_\_.py" to "main.py"**
2. **Copy all content from \_\_init\_\_.py to main.py**
3. **Clear all content in \_\_init\_\_.py**

### Contents of the requirements.txt file:
azure-functions

### Related information

* Links to source
* Bindings used

**Comment by lzchen** [comment]
@lzchen

**Comment by vrdmr** [comment]
Happens for 3.8 as well.
```
➜ python --version
Python 3.8.6
```

I added the following in my simple HTTP trigger function
```python
print("+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=Hello World")
print(traceback.print_stack(file=sys.stdout))
```

And got the following stacktrace twice. We need to confirm if the host itself is sending the load call twice when being invoked.
```python
+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=Hello World
  File "/Users/varadmeru/work/microsoft/azfunctions/pyworker/worker.py", line 19, in <module>
    main.main()
  File "/Users/varadmeru/work/microsoft/azfunctions/pyworker/azure_functions_worker/main.py", line 47, in main
    return aio_compat.run(start_async(
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/varadmeru/work/microsoft/azfunctions/pyworker/azure_functions_worker/dispatcher.py", line 244, in _dispatch_grpc_request
    resp = await request_handler(request)
  File "/Users/varadmeru/work/microsoft/azfunctions/pyworker/azure_functions_worker/dispatcher.py", line 273, in _handle__function_load_request
    func = loader.load_function(
  File "/Users/varadmeru/work/microsoft/azfunctions/pyworker/azure_functions_worker/utils/wrappers.py", line 32, in call
    return func(*args, **kwargs)
  File "/Users/varadmeru/work/microsoft/azfunctions/pyworker/azure_functions_worker/loader.py", line 76, in load_function
    mod = importlib.import_module(fullmodname)
  File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/varadmeru/work/microsoft/stencil/StencilDemo/test-vscode-azfunc1/HttpTrigger1/__init__.py", line 9, in <module>
    print(traceback.print_stack(file=sys.stdout))
```

Interestingly, the following log statements come in, but the prints do not happen in there.
```log
[2020-10-19T06:59:40.643] Worker process started and initialized.
[2020-10-19T06:59:40.648] Received FunctionLoadRequest, request ID: 6b260f3a-0a38-49d5-98a9-0312896e779b, function ID: 251ccd85-63dc-4830-b5c6-c7b74ca9103c
[2020-10-19T06:59:40.704] Successfully processed FunctionLoadRequest, request ID: 6b260f3a-0a38-49d5-98a9-0312896e779b, function ID: 251ccd85-63dc-4830-b5c6-c7b74ca9103c
```

**Comment by AnatoliB** [comment]
@Hazhzeng, @vrdmr This may be related: https://github.com/Azure/azure-functions-powershell-worker/issues/276

**Comment by Hazhzeng** [comment]
Thanks as I'm still investigating what's the actual issue might be.
1. The host only send FunctionLoadRequest once
2. The worker only receive and process FunctionLoadRequest once
3. The `mod = importlib.import_module(fullmodname)` statement is executed twice with **only one load_function** call

```
Established RPC channel. WorkerId: 83e29d78-73ba-418e-875b-7242b03cfd0c
Worker Process started. Received StartStream message
Switched to gRPC logging.
Received WorkerInitRequest, request ID e4425a1d-4691-4507-ba56-12044e653175
Received WorkerInitResponse. Worker process initialized
Worker capabilities: { "RawHttpBodyBytes": "true", "TypedDataCollection": "true", "RpcHttpBodyOnly": "true", "RpcHttpTriggerMetadataRemoved": "true" }
Updating capabilities: { "RawHttpBodyBytes": "true", "TypedDataCollection": "true", "RpcHttpBodyOnly": "true", "RpcHttpTriggerMetadataRemoved": "true" }
Adding jobhost language worker channel for runtime: python. workerId:83e29d78-73ba-418e-875b-7242b03cfd0c
Sending FunctionLoadRequest for function:HttpTrigger with functionId:cd7e9fe8-8d5b-46a8-a86f-2ce58495092c
Worker process started and initialized.
Received FunctionLoadRequest, request ID: e4425a1d-4691-4507-ba56-12044e653175, function ID: cd7e9fe8-8d5b-46a8-a86f-2ce58495092c
Host lock lease acquired by instance ID '00000000000000000000000055FF78A3'.
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
File "c:\Users\hazeng\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy\__main__.py", line 45, in <module>
    cli.main()
File "c:\Users\hazeng\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 444, in main
    run()
File "c:\Users\hazeng\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 285, in run_file
    runpy.run_path(target_as_str, run_name=compat.force_str("__main__"))
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64/worker.py", line 86, in <module>
    main.main()
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\main.py", line 47, in main
    return aio_compat.run(start_async(
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\base_events.py", line 629, in run_until_complete
    self.run_forever()
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\windows_events.py", line 316, in run_forever
    super().run_forever()
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\base_events.py", line 596, in run_forever
    self._run_once()
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\base_events.py", line 1890, in _run_once
    handle._run()
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\dispatcher.py", line 242, in _dispatch_grpc_request
    resp = await request_handler(request)
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\dispatcher.py", line 271, in _handle__function_load_request
    func = loader.load_function(
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\utils\wrappers.py", line 32, in call
    return func(*args, **kwargs)
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\loader.py", line 76, in load_function
    mod = importlib.import_module(fullmodname)
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
File "C:\Users\hazeng\Projects\Functions\PythonNoReq\HttpTrigger\__init__.py", line 10, in <module>
    invoke()
File "C:\Users\hazeng\Projects\Functions\PythonNoReq\OtherModule\__init__.py", line 7, in invoke
    for line in traceback.format_stack():
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
File "c:\Users\hazeng\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy\__main__.py", line 45, in <module>
    cli.main()
File "c:\Users\hazeng\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 444, in main
    run()
File "c:\Users\hazeng\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 285, in run_file
    runpy.run_path(target_as_str, run_name=compat.force_str("__main__"))
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64/worker.py", line 86, in <module>
    main.main()
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\main.py", line 47, in main
    return aio_compat.run(start_async(
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\base_events.py", line 629, in run_until_complete
    self.run_forever()
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\windows_events.py", line 316, in run_forever
    super().run_forever()
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\base_events.py", line 596, in run_forever
    self._run_once()
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\base_events.py", line 1890, in _run_once
    handle._run()
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\dispatcher.py", line 242, in _dispatch_grpc_request   
    resp = await request_handler(request)
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\dispatcher.py", line 271, in _handle__function_load_request                                                                                                                                                                                                                        quest
    func = loader.load_function(
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\utils\wrappers.py", line 32, in call
    return func(*args, **kwargs)
File "C:\Users\hazeng\AppData\Roaming\npm\node_modules\azure-functions-core-tools\bin\workers\python\3.9/WINDOWS/X64\azure_functions_worker\loader.py", line 76, in load_function
    mod = importlib.import_module(fullmodname)
File "C:\Users\hazeng\AppData\Local\Programs\Python\Python39\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
Successfully processed FunctionLoadRequest, request ID: e4425a1d-4691-4507-ba56-12044e653175, function ID: cd7e9fe8-8d5b-46a8-a86f-2ce58495092c
Received FunctionLoadResponse for functionId:cd7e9fe8-8d5b-46a8-a86f-2ce58495092c
``` 

Checking if this is related to the virtual namespace \_\_app\_\_ issue.

**Comment by Hazhzeng** [comment]
OK, so if a customer doesn't write their function code in \_\_init\_\_.py, but change the scriptFile to main.py, this would not be an issue.

**Comment by Hazhzeng** [comment]
Hi @lzchen sorry for the long investigation. The mitigation is provided in the description, just need to change \_\_init\_\_.py filename to something else (e.g. main.py). Also change the function.json => scriptFile to match it.

**Comment by lzchen** [comment]
@Hazhzeng 
Thanks for looking into this.
Has this behaviour always been like this on load? How come customers have never run into this before?

**Comment by Hazhzeng** [comment]
@lzhen, to be honest, I'm also pretty surprise by this behavior. This should be a very old issue. I checkout the 1.1.5 tag, this issue was still there.

https://github.com/Azure/azure-functions-python-worker/blob/dev/azure_functions_worker/loader.py#L70-L72
The loader function does not have any special cases to distinguish loading the package initialization \_\_init\_\_.py file. Making a fix this it now.

**Comment by mathrb** [comment]
Hello,

I faced the same issue today and ended up googling this double loading.
As mentioned by @Hazhzeng, I changed my script from `__init__.py` to `main.py`, updated the scriptFile in the function.json file and now my code outside main is loaded only once.
What surprise me is:
1. It seems a fix was published Jan 27th 2021 #811 
2. The label of this tag is in-progress

`func --version` -> `4.0.4483`
`python --version` -> `Python 3.8.10`

Kind regards
